File: python/797_AllPathsFromSourceToTarget.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        res = []
        def dfs(idx, curr):
            for node in graph[idx]:
                dfs(node, curr+[node])
            
            if idx == len(graph)-1:
                res.append(curr)
        
        dfs(0, [0])
            
        return res

File: java/797_AllPathsFromSourceToTarget.java
class Solution {
    private List<List<Integer>> res = new ArrayList<List<Integer>>();
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<Integer> start = new ArrayList<Integer>();
        start.add(0);
        dfs(0, start, graph);
        return this.res;
    }
    public void dfs(int idx, List<Integer> currList, int[][] graph) {
        for (int node : graph[idx]) {
            currList.add(node);
            dfs(node, currList, graph);
            currList.remove(currList.size()-1);
        }
        if (idx == graph.length-1) {
            this.res.add(new ArrayList<Integer>(currList));
            return;
        }
    }
}
